mod_s3_upload/src/mod_s3_upload.erl

```diff
@@ -0,0 +1,462 @@
+            auth         = Auth,
+            access       = get_opt(access, Opts)}.
+
+-spec url_service_parameters(
+        Params :: #params{}
+       ) ->
+          ServiceParameters :: [{binary(), binary() | true}].
+% additional URL parameters from module config
+url_service_parameters(#params{set_public = true}) ->
+    [{<<"X-Amz-Acl">>, <<"public-read">>}];
+url_service_parameters(_) ->
+    [].
+
+-spec upload_parameters(
+        UploadRequest :: #upload_request_0{},
+        Params :: #params{}
+       ) ->
+          UploadParameters :: [{binary(), binary() | true}].
+% headers to be included with the PUT request
+upload_parameters(#upload_request_0{size           = FileSize,
+                                    'content-type' = ContentType},
+                  ServiceParams) ->
+    [{<<"Content-Type">>,   <<ContentType/binary>>},
+     {<<"Content-Length">>, erlang:integer_to_binary(FileSize)}
+    | url_service_parameters(ServiceParams)].
+
+-spec put_url(
+        UploadRequest :: #upload_request_0{},
+        Params :: #params{},
+        URL :: binary()
+       ) ->
+          PutURL :: binary().
+% attach additional query parameters (to the PUT URL), specifically canned ACL.
+put_url(UploadRequest, ServiceParams, URL) ->
+    UriMap = uri_string:parse(URL),
+    QueryList = case UriMap of
+                    #{query := QueryString} ->
+                        uri_string:dissect_query(QueryString);
+                    _ ->
+                        []
+                end,
+    Params = upload_parameters(UploadRequest, ServiceParams),
+    WithOpts = uri_string:compose_query(Params ++ QueryList),
+    uri_string:recompose(UriMap#{query => WithOpts}).
+
+-spec object_url(
+        BucketURL :: binary(),
+        FileName :: binary()
+       ) ->
+          ObjectURL :: binary().
+% generate a unique random object URL for the given filename
+object_url(BucketURL, FileName) ->
+    #{path := BasePath} = UriMap = uri_string:parse(BucketURL),
+    ObjectName = object_name(FileName),
+    uri_string:recompose(UriMap#{path => <<BasePath/binary, "/", ObjectName/binary>>}).
+
+-spec object_name(
+        FileName :: binary()
+       ) ->
+          ObjectName :: binary().
+% generate a unique-in-time object name. the name consists of a hash
+% derived from the file, node, time, and a random number, a
+% forward-slash, and the original filename. this ensures that it does
+% not collide with other objects, while the forward slash ensures that
+% the client displays only the original file name.
+object_name(FileName) ->
+    MD = crypto:hash_init(sha256),
+    MDFilename = crypto:hash_update(MD, FileName),
+    MDNodeName = crypto:hash_update(MDFilename, atom_to_list(node())),
+    MDTime = crypto:hash_update(MDNodeName, <<(os:system_time())>>),
+    MDRand = crypto:hash_update(MDTime, crypto:strong_rand_bytes(256)),
+    Hash = crypto:hash_final(MDRand),
+    <<(str:to_hexlist(Hash))/binary, "/", FileName/binary>>.
```

Comment:
I think I may modify this to make less of a management nightmare for small operators. Although cloud storage objects do not technically have a hierarchy within a bucket and all live at the top level, many tools like the AWS CLI will still display them as if they are in folders. Obviously this will lead to there being a lot of synthetic folders.

I'll address this tomorrow. It's rather late here as of this comment.